Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+
+### Lambda
+
+We also add an enum to store a static function and a property on. We put this on the static `Lambda` because `LambdaRuntime` is generic and thus has bad ergonomics for static properties and functions.
+
+```swift
+enum Lambda {
+    /// This returns the default EventLoop that a LambdaRuntime is scheduled on.
+    /// It uses `NIOSingletons.posixEventLoopGroup.next()` under the hood.
+    public static var defaultEventLoop: any EventLoop { get }
+
+    /// Report a startup error to the Lambda Control Plane API
+    public static func reportStartupError(any Error) async
+}
+```
+
+Since the library now provides ownership of the `main()` function and allows users to initialize services before the `LambdaRuntime` is initialized, the library cannot implicitly report [errors that occur during initialization to the dedicated endpoint AWS exposes](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html#runtimes-api-initerror) like it currently does through the `initialize()` function of `LambdaRunner` which wraps the handler’s `init(...)` and handles any errors thrown by reporting it to the dedicated AWS endpoint.
+
+To retain support for initialization error reporting, the `Lambda.reportStartupError(any Error)` function gives users the option to manually report initialization errors in their closure handler. Although this should ideally happen implicitly like it currently does in v1, we believe this is a small compromise in comparison to the benefits gained in now being able to cleanly manage the lifecycles of required services in a structured concurrency manner.
+
+
+>Use-case:
+>
+>Assume we want to load a secret for the Lambda function from a secret vault first.
+>If this fails, we want to report the error to the control plane:
+>```swift
+>let secretVault = SecretVault()
+>
+>do {
+>    /// !!! Error thrown: secret "foo" does not exist !!!
+>    let secret = try await secretVault.getSecret("foo")
+>
+>    let runtime = LambdaRuntime { (event: Input, context: LambdaContext) in
+>        /// Lambda business logic
+>    }
+>
+>    let serviceGroup = ServiceGroup(
+>        services: [postgresClient, runtime],
+>        configuration: .init(gracefulShutdownSignals: [.sigterm]),
+>        logger: logger
+>    )
+>    try await serviceGroup.run()
+>} catch {
+>    /// Report startup error straight away to the dedicated initialization error endpoint
+>    try await Lambda.reportStartupError(error)
+>}
+>```
+
+### Codable support
+
+For v2 we want to introduce a `LambdaCodableAdapter` that:
+
+1. Accepts any generic underlying handler conforming to a new protocol `LambdaHandler` specifying `associatedtype`s `Event` and `Output`, and a `handle(_ event: Event, context: LambdaContext) async throws -> Output` function.
+
+#### LambdaHandler
+
+```swift
+public protocol LambdaHandler {
```

Comment:
Just out of curiosity we can't solve this with protocol inheritance right? We do need the various different protocols and adapters?

Reply:
We do have protocol inheritance today and I don't like it. The problem here is that there is never a case where the protocol inheritance here makes sense from a user point of view. You either want to use pure streaming or you want to use Codable. But there is no point in which you opt to override the streaming API when you implement `LambdaHandler`. Worst of all injecting custom Encoders and Decoders only works through globals.